// typedef/Gruntfile.js
module.exports = function(grunt) {
    'use strict';

    // Project configuration.
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),

        jshint: {
            lib: ['lib/**/*.js'],
            test: ['test/main.js', 'test/unit/*.js'],
            grunt: ['Gruntfile.js']
        },

        browserify: {
            debug: {
                src: ['lib/<%= pkg.name %>.js'],
                dest: 'bin/<%= pkg.name %>.debug.js',
                options: { debug: true }
            },

            release: {
                src: ['lib/<%= pkg.name %>.js'],
                dest: 'bin/<%= pkg.name %>.release.js',
                options: { standalone: '<%= pkg.name %>' }
            },

            releaseNoVendor: {
                src: ['lib/<%= pkg.name %>.js'],
                dest: 'bin/<%= pkg.name %>.release.no-vendor.js',
                options: { standalone: '<%= pkg.name %>', ignore: 'underscore' }
            },

            test: {
                src: ['test/main.js'],
                dest: 'test/bin/main.debug.js',
                options: { debug: true }
            }
        },

        uglify: {
            release: {
                src: ['bin/<%= pkg.name %>.release.js'],
                dest: 'bin/<%= pkg.name %>.min.js'
            },
            releaseNoVendor: {
                src: ['bin/<%= pkg.name %>.release.no-vendor.js'],
                dest: 'bin/<%= pkg.name %>.no-vendor.min.js'
            },
            options: { report: 'gzip' }
        },

        watch: {
            all: {
                files: [
                    'lib/**/*.js',
                    'test/main.js',
                    'test/unit/**/*.js'
                ],
                tasks: ['default']
            }
        },

        qunit: {
            all: ['test/index.html']
        }

    });

    // plugins
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-browserify');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-qunit');

    // tasks
    grunt.registerTask('lint', ['jshint']);
    grunt.registerTask('build', ['lint', 'browserify']);
    grunt.registerTask('test', ['lint', 'browserify:test', 'qunit']);
    grunt.registerTask('default', ['build', 'uglify']);

};
